CRM/Admin/Page/AJAX.php

```diff
@@ -304,20 +304,24 @@ public static function getTagTree() {
     $parent = CRM_Utils_Type::escape(CRM_Utils_Array::value('parent_id', $_GET, 0), 'Integer');
     $result = array();
 
-    $parentClause = $parent ? "AND tag.parent_id = $parent" : 'AND tag.parent_id IS NULL';
-    $sql = "SELECT tag.*, child.id AS child, COUNT(et.id) as usages
-      FROM civicrm_tag tag
-      LEFT JOIN civicrm_entity_tag et ON et.tag_id = tag.id
-      LEFT JOIN civicrm_tag child ON child.parent_id = tag.id
-      WHERE tag.is_tagset <> 1 $parentClause
-      GROUP BY tag.id
-      ORDER BY tag.name";
+    $parentClause = $parent ? "AND parent_id = $parent" : 'AND parent_id IS NULL';
+    $sql = "SELECT *
+      FROM civicrm_tag
+      WHERE is_tagset <> 1 $parentClause
+      GROUP BY id
+      ORDER BY name";
+
+    // fetch all child tags in Array('parent_tag' => array('child_tag_1', 'child_tag_2', ...)) format
+    $childTagIDs = CRM_Core_BAO_Tag::getChildTags();
```

Comment:
Is this absolutely necessary? If we don't need the usage data can we re-add the `LEFT JOIN civicrm_tag child ON child.parent_id = tag.id` back to the query? That seems to be the only purpose of loading this massive array into memory.

Reply:
Problem with the left join is that it doesn't look deep down the hierarchy to retrieve level-N child tags, reason why I need to introduce this new fn to fetch the tag tree via recursive search. However, we can avoid this lookup on each call by storing the  ```$childTagIDs``` array in a ```Civi::static``` variable? And need to ensure that whenever a tag got deleted/moved in the tag tree this static variable got repopulated.

Reply:
This function, by design, does not need to look deep within the hierarchy. It only needs to know whether the tag has children or not. 1 level deep is fine.